### ISSUE TITLE
Instancing GPUParticles3D Causes Flashes Onscreen

### ISSUE THREAD

**Issue by The-Randalorian** [godotengine/godot#71935]
### Godot version

v4.0.beta14.official [28a24639c]
v4.0.1.stable.official [cacf49999]

### System information

Windows 11 22H2, Forward+, NVIDIA MX350 (528.02) and NVIDIA RTX2070 (517.00)
Ubuntu 22.04, Forward+, NVIDIA MX350 (525.85.05)

### Issue description

Instancing a GPUParticles3D can cause colored flashes on screen. In some cases (such as the minimal reproduction project), the flashes are small, however in my other project I have seen them fill the screen. I am still unsure what causes the different sizes

These are some example correct frames from the minimal reproduction project (the effect is supposed to be a fireball.)

<img width="578" alt="image" src="https://user-images.githubusercontent.com/8741983/214133602-ff0e5916-fa0b-4e52-8dac-9701b8220a1d.png">

<img width="578" alt="image" src="https://user-images.githubusercontent.com/8741983/214133646-c976fde1-569c-4d0a-9395-fd037f21e0b9.png">

These are some example incorrect frames.

<img width="578" alt="image" src="https://user-images.githubusercontent.com/8741983/214133760-8e4804e8-f194-4d31-b943-0d93ab6d60cb.png">

<img width="578" alt="image" src="https://user-images.githubusercontent.com/8741983/214132553-c6d4be93-5373-43b5-b52e-d9a33924e9bf.png">

### Steps to reproduce

Create two scenes, one with a GPUParticles3D, and one that instances those particles repeatedly, using a script. Flashes will occur on screen.

### Minimal reproduction project

[Particle Flash MRP.zip](https://github.com/godotengine/godot/files/10483071/Particle.Flash.MRP.zip)

Edit 1: Corrected driver version for my friend's computer.
Edit 2: Added note about the issue also appearing on Linux, and that the issue is still present in 4.0.1

*Bugsquad edit (keywords for easier searching): flicker, flickering*

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/70546.

**Comment by clayjohn** [comment]
Might be fixed already by https://github.com/godotengine/godot/pull/71925

Could you test with https://github.com/godotengine/godot/pull/71925 and see if it fixes the issue for you?

**Comment by clayjohn** [comment]
No problem, https://github.com/godotengine/godot/pull/71925 was just merged, so you can test again when Beta 15 releases in the next couple of days!

**Comment by The-Randalorian** [comment]
Just tested it in v4.0.beta.custom_build [379d83dd5], which had #71925. The issue persisted.

I am attempting another build with some of the newer commits (up to 70377), to verify that I didn't mess up something with git, but I'm pretty sure I did everything correctly.

**Comment by The-Randalorian** [comment]
I should note the flashes that appear seem to be the particle mesh; the round shape is due to the radial gradient used for the flame effect. Using a noise map yields a result like this (only the first picture is correct.)

<img width="578" alt="image" src="https://user-images.githubusercontent.com/8741983/214159785-8c583e58-5964-49d4-9472-5998f37f9cf7.png">

<img width="578" alt="image" src="https://user-images.githubusercontent.com/8741983/214158342-75ed7916-63f1-4e3e-bb68-0db26e0a3d8b.png">

<img width="578" alt="image" src="https://user-images.githubusercontent.com/8741983/214160025-8f430368-f819-4150-afa7-31c6b5ff6167.png">

**Comment by IvanIG3** [comment]
I'm also having this issue in 4.0.stable. It happens almost always, when I create a new instance of a GPUParticles3D. In the video below, 3 of the 4 instances generates a big colored flash.

![flash0](https://user-images.githubusercontent.com/54667250/225355606-4459eb06-2d43-41d3-b1b1-26d3b9924be0.png)

![flash1](https://user-images.githubusercontent.com/54667250/225355262-fbe60f55-a456-4410-9e81-90693e071796.png)

https://user-images.githubusercontent.com/54667250/225355772-3606b90e-a8e2-4d12-95a8-887efd929ff2.mp4

**Comment by The-Randalorian** [comment]
The issue is not limited to just windows, also appearing on my laptop after I swapped it to Linux. I have added a note to the original issue.

**Comment by hunterloftis** [comment]
I'm experiencing the same thing with Godot_v4.0-stable_linux.x86_64 on Fedora 37. It's intermittent, but always happens a few times:


https://user-images.githubusercontent.com/364501/228417424-f6901ca8-743f-4f11-bb2a-00d810e4f1e1.mp4

**Comment by Calinou** [comment]
I can confirm this on 4.0.1.stable (Linux, GeForce RTX 4090 with NVIDIA 525.89.02).

**Epilepsy warning: The videos below contain flashing.**

### Default MRP settings

https://user-images.githubusercontent.com/180032/228615518-3a8255c8-87e2-4796-8080-e57e739594d6.mp4

### Interpolation disabled, Fixed FPS = 0 (use render FPS)

https://user-images.githubusercontent.com/180032/228615522-7d14dbe0-100c-462a-9bc5-629a2c207350.mp4

### Interpolation disabled, Fixed FPS = 1

https://user-images.githubusercontent.com/180032/228615526-11c189aa-a983-49b5-a337-823233bb568e.mp4

### Preprocess = 0.25s (same as Lifetime)

https://user-images.githubusercontent.com/180032/228615527-8f6634fb-ae22-4428-bb0d-baec0067571c.mp4

**Comment by ParasiteSt** [comment]
I got a link to this bug after posting about this issue. I can confirm it is a bug in v4.0.2 . I will watch this bug report going forward, thanks Calinou for the link to this post!

**Comment by ParasiteSt** [comment]
I hope this can be looked into further before v4.1 release, while it can be totally worked around and is only visually impacted, this bug really hurts the particle system as a whole in Godot Vulkan.

**Comment by ParasiteSt** [comment]
> What do you do to workaround this issue?

Oh sorry, I meant it like I just deal with the issue and you can work around it being there, its not engine breaking or anything. It may not usable if you are prone to epileptic seizures though.

**Comment by The-Randalorian** [comment]
> > I hope this can be looked into further before v4.1 release, while it can be totally worked around and is only visually impacted, this bug really hurts the particle system as a whole in Godot Vulkan.
> 
> What do you do to workaround this issue?

I've had limited success through moving the rest of the game world away from the origin, but it's not a consistent or reliable fix. Sometimes the particles appear HUGE, and will therefore still cause flashes even if the level is a decent distance from the origin. Trying to move the level further can yield physics and rendering bugs as the floats get less precise.

**Comment by RichardR01** [comment]
Godot 4.02, and 4.03 rc2:
I am having similar issues. I have an animation, where 12 missiles are being instanced. The missile has 2 GPUParticles3D to create the effect of flame and smoke. They start being turned off. When the player presses the "fire" button, it enables the physics and turns on the particles after a brief timer timeout. The player can fire consecutively, firing all 12 missiles. The player can also reload, adding even more. The problem doesn't surface the first time, but after the 2nd or 3rd reload, I am seeing flashes of white on the right hand side of the screen, and particles being drawn in places they don't belong.

**Comment by RichardR01** [comment]
I found something that helps. Keep the emitting on, set the visible property to false. Set the preprocess to 0.1. I also set the visibility to false in the exit function. That way when the particles are instantiated, they are running but not visible, with the 0.1 preprocess buffer.

**Comment by IvanIG3** [comment]
Hi! What I'm doing to bypass this issue (kinda), is: instantiate the particles with transparency, wait until it stops flashing, and then restart the particles and remove the transparency. The code looks like this:

```
# "particles" is an instance of GPUParticles3D
particles.transparency = 1.0
particles.emitting = true
await get_tree().create_timer(0.2).timeout
particles.restart()
particles.emitting = true # this again or it freezes
particles.transparency = 0.0
```

It does not always works, but it helps. The big problem of this workaround is the delay for displaying the particles.

**Comment by clayjohn** [comment]
This looks to be an issue specifically when ``draw_order`` is set to "view depth". Looking at the particles buffers. It seems the positions get out of sync in that case. I'm still investigating where the issue comes from, but this should be fixable in a safe way for 4.2 and 4.1.1

**Comment by IvanIG3** [comment]
> This looks to be an issue specifically when `draw_order` is set to "view depth". Looking at the particles buffers. It seems the positions get out of sync in that case. I'm still investigating where the issue comes from, but this should be fixable in a safe way for 4.2 and 4.1.1

I can reproduce this bug with any draw_order setting. Maybe I'm wrong but I don't think this is exclusive for "view depth".

**Comment by clayjohn** [comment]
@IvanIG3 You can reproduce the issue with all draw_order settings with the MRP from this issue, or in your own project?

**Comment by clayjohn** [comment]
I've made a PR that fixes the MRP from this issue. https://github.com/godotengine/godot/pull/78852

@IvanIG3 I would appreciate if you could test https://github.com/godotengine/godot/pull/78852 and confirm whether it fixes your project as well

**Comment by IvanIG3** [comment]
> I've made a PR that fixes the MRP from this issue. #78852
> 
> @IvanIG3 I would appreciate if you could test #78852 and confirm whether it fixes your project as well

@clayjohn I don't know how to do that. I made a MRP, extracted directly from my project. Could you test it? You only need to run the project and click the "play" button. The bug triggers 50% of the time. If it doesn't with a few clicks, then I think it's safe to say that the bug is fixed. Thanks!

[GPUParticlesBug.zip](https://github.com/godotengine/godot/files/11915353/GPUParticlesBug.zip)

**Comment by clayjohn** [comment]
@IvanIG3 https://github.com/godotengine/godot/pull/78852 fixes your MRP as well :) 

You can expect the fix in 4.2 and in 4.1.1 (it is a little bit too late for 4.1, but 4.1.1 should only be a few weeks behind)
